Make HTML semantically correct

retrocausal opened the pull request:
- The binding element is main
 - Our App is an article
 - App will have sections and divs
 - Also, lowercase css classes in accordance with Udacity best practices

The changes here are in accordance with maintaining semantically correct HTML tags.
though this seems petty, having semantically correct tagging on the DOM, enables the markup to be more readable and index-able, as opposed to a div soup

Also, css classes are best written in lowercase.

overall, is a reflection of [Udacity's style guide](https://udacity.github.io/frontend-nanodegree-styleguide/)

retrocausal committed:
Make HTML semantically correct
 - The binding element is main
 - Our App is an article
 - App will have sections and divs
 - Also, lowercase css classes in accordance with Udacity best practices

skyshader commented on client/src/App/App.js:
```diff
@@ -3,7 +3,10 @@ import PropTypes from "prop-types";
+  //Make the binding lement semantically correct
+  //Our app is an article, and will have <sections> and <divs>
+  //as children
+  return <article className="app">{children}</article>;
```
App is not an article I think. Did you consider using `section`?

retrocausal replied:
Sure, create-react-app only bootstraps what is necessarily react dependent to perfection.
IDK if they emphasize on semantic parameters. 
I believe, App itself is the first content /child of the `root` identifier. 

so the `root` becomes `main` and App becomes `article`  - > meaning, an article can have sections and divs. 
A section having an article sounds improper IMO

twishasaraiya replied:
@retrocausal We should definitely maintain HTML semantics. Css style guide makes sense but IMOi think we  should go with `div` Since `article` is that is independent.
`The HTML <article> element represents a self-contained composition in a document, page, application, or site, which is intended to be independently distributable or reusable `

retrocausal replied:
thanks @twishasaraiya ! I can change the App element to be a `<div>`. and resubmit the PR. 
does this look good?

`main:root` - > `div:app` -> `div:home`->`section:header`->`section:greensock`->`section:footer` ? 

We do not have the components forming header/greensock/footer yet, but they will need to be `sections` .

Suggestion: 
can we rename the 'home' component? for now since there has been no development on the client side yet, it makes sense to get rid of the redundant `div` inside `App` ? or rename it to header and make it a component instead of a page? 

I am guessing it is better we get the rendering containers organised before we dump data /animations into them

retrocausal replied:
hmm.. will look into creating components in a bit, and maybe create the four we require header nav footer and greensock. And will get back to you. can we freeze PRs till we have our containers ready? Will speed it up from my end, and may submit all changes by friday

skyshader reviewed (commented):
Btw, This was as per what `create-react-app` generated. So I considered we should follow that. But yeah your points are valid too. So let's see how well can we maintain the semantics.

anantoghosh commented on client/src/pages/Home/Home.js:
```diff
@@ -8,7 +8,9 @@ class Home extends PureComponent {
-      <div className="Home container">
+      //ensure all css classes are lowercase - best practices
+      //this is the container for all things within the article App
+      <div className="home container">
```
With CSS modules we cannot use lowercase naming convention.
Since these names are used inside Javascript where "`-` "is not allowed.
naming has to be in the form `flatButton` or `FlatButton`.
But we cannot use `flat-button`. Using no spectator for words is **not** preferred eg `flatbutton`.

anantoghosh replied:
We should be using css modules in Home.js.
The index.css will be the only global css file correct?

skyshader replied:
Yes correct. And we are using separate css files for each component. But not using `css-modules` per se. It's just like writing normal css.

anantoghosh replied:
oh I see, I was under the impression that cra was pre configured with css modules.
Then lowercase naming is fine by me.

skyshader replied:
It is sort of pre configured with css-modules but uses only autoprefixer.

Sure, shall we merge this PR then? If all approves.

retrocausal replied:
I did not know about the naimng conventions within react per se. 
However, If it is only one single word, it can be lowercase. 
the issue you might be referring to @anantoghosh , might happen, only when using two or more words in a name. In which case, we can always have camel / snake cases as react expects it to be in case we use a module.

skyshader replied:
So, to overcome the issue, I used to use `_`s instead of `-`s

skyshader replied:
And by the way there are many different style guides out there to name css classes. It's not fixed to one being referenced here. So I think we can stick with what we are following.
`.ComponentName_heading` and `.ComponentName_container` and stuff life that.

There is a reason behind what we are following: We avoid overlaps in name. Suppose we want to have .container class in two component which would overlap. Prefixing with component's name would resolve that.

I know there are lots of other ways css-modules adds a hash to the classname to prevent this and other solutions but we are following this and I think we should stick with this.

Do let me know your thoughts: @retrocausal and @anantoghosh

retrocausal commented on client/public/index.html:
```diff
@@ -41,7 +41,7 @@
       You need to enable JavaScript to run this app.
     </noscript>
-    <div id="root"></div>
+    <main id="root"></main>
```
we aren't :) react dumps everything inside <main>. Hence, there is no logical seperation of concerns when it comes to markup. 

Either we hard code a <header> and a <footer>, or, we create a header component and a footer component
these will have to be <sections> with classes header and footer. 

I am only changing the tag name to main, so that it is semantic in nature, and not just another div

anantoghosh replied:
We will have a header, footer, navigation components and that should have the `header` , `footer`, `nav` tags etc.

retrocausal replied:
Cool. And how would react render them? will it create tags adjacent the tag with the id `root`? 
or will they be children? 
If they are children, they still need to be sections? with classnames header and nav and footer?
Asking as I hve literally no idea how react adds components

skyshader replied:
Yes they are going to be children. The `header`, `footer`, `nav` elements. See what fit's best as per your thought process. I will appreciate your judgement here.

retrocausal committed:
Merge branch 'udevelop' into develop

retrocausal committed:
Add Initial UI and make it responsive
 - Create three components
    - Header
    - Footer and
    - Greensock
 - Create Containers for the above new components
 - Add Immutable CSS classes for maintaining UI responsiveness
 - Add Visually identifiable css properties (THIS IS ONLY TILL WE HAVE COMPONENT LEVEL CODE)
    - We need to remove those borders once we develop these individual sections

skyshader commented on client/src/components/FooterComponent/FooterComponent.js:
```diff
@@ -0,0 +1,12 @@
+import React from "react";
+import PropTypes from "prop-types";
+
+const FooterComponent = ({ message }) => {
```
Let's name it Footer. It's redundant as it is already placed inside components directory.

retrocausal replied:
Done

skyshader commented on client/src/components/FooterComponent/FooterComponent.js:
```diff
@@ -0,0 +1,12 @@
+import PropTypes from "prop-types";
+
+const FooterComponent = ({ message }) => {
+  return <section className="Footer_component">{message}</section>;
```
Footer_component : NO
Footer_container: YES

retrocausal replied:
Done

retrocausal replied:
regarding if we need a footer, which is not in the design mockup you shared, I included a footer because, that is generally a good place to link back to the repo. and more importantly, 

 - We absolutely need to _credit_ and _attribute_ the sources from where we got our questions, and also the attributions for any images we may require. 

Hence, a footer is required to include these considerations.

anantoghosh replied:
`<footer>` tag for footer.

skyshader commented on client/src/components/GreensockComponent/GreensockComponent.js:
```diff
@@ -0,0 +1,12 @@
+import React from "react";
+import PropTypes from "prop-types";
+
+const GreensockComponent = ({ message }) => {
```
Let's remove this Greensock component for now. It's not required IMO. We will be having component that will use Greensock and not a Greensock component itself.

retrocausal replied:
So, the current setup is, 
Route: `home`
      Page: `home`
          Containers: `Header` and `Footer` and `Greensock`
Each container defines the render properties, default or otherwise
Each container then, passes these properties to its associated component.

The component associated with the container, renders the content.

So we need three components, for each of the three containers. 
  - Containers prep data,
  - Accompanying Components wrap data inside markup and render.

That was how, the sample container was used by the page home which rendered "This is a sample page" on screen.

I am unsure of how greensock itself works At the moment. I am assuming, 

 - We get an animatable object which in turn has 
    - the quote
    - the question
    - the list of answers

these need a container to fit into. 
the greensock container, will prep the data as required, and make them renderable properties
the greensock container then binds this renderable property to it's component.
the greensock component defines its animations
the greensock component then renders animated content.

the current greensock component, is only a placeholder, and is not coded. Not even a single line of js. 

maybe we can keep it that way, so when we need to write the associated JS, we have a placeholder component to edit?

twishasaraiya replied:
GreenSock has very simple syntax such as 
TweenMax(#id, duration, {css property object}) that's it and the `div` with `id` gets animated. This can be done in `componnetDidMount` method of the component. IMO separate greensock container is not required.

anantoghosh replied:
Are you saying that GreenSock container will wrap every animatable object?

skyshader replied:
I think we should use general Content container and add components inside it. The components can use greensock api to provide animations.

retrocausal replied:
To be clear, What I am saying, this: 

1. We need 4 Sections (Header+Footer+Content+Nav) (*Everybody agrees?*) (shaan has reservations for the footer)

2. Now, If we have a footer, we can attribute our sources there, and have our link backs to the repo

3. I do not know yet, what forms the header, we can decide that later. We now just need a Empty section for the header.

4. Between the header and the footer Sits the **content** . This Section, will have 

    a. The Quote, 
    b. The Answers
    c. Anything else necessary

5. If the above is agreed upon, Then, **Where** are we getting the content from ? From a data API(either our won / third party)

6. Where and How do we display these quotes and their paraphernalia ?

     a. We prepare the **Render Props** in the **Content Container** , and then pass them to the **Content Component**
     b. If as @twishasaraiya  mentions, Greensock is easy to setup, It needs a DOM ID. 
     c. Keeping (b) in mind, this DOM element, Must have rendered our content from (5)

And so, we need the content container, bound to  the content component
All with me? I am coming to why I chose to do it that way in a bit. 

### Now, How Is the **Initial** **React generated code** laid out? 

1. There is a `PAGE`

2. A `PAGE` has a `CONTAINER` (`sampleContainer`)

3. The SampleContainer , **defines** all **Properties** that represent the data(**Model**)

4. The sampleContainer then, **Binds** to a **Component** the `sampleComponent` and passes renderable props. 

**The above** is how, react does it. 

### What we are doing in this PR ?

1. We are **copying** how react itself generates an app. 

2. We are copying react's `sampleContainer` to three copies (Header/Footer/Content). We can always add Nav. 

3. We are then copying react's `sampleComponent` to three copies (Header/Footer/Content). 

And we are changing **names**. That is all :) 

We are not **coding** a component. 

I guess, all that coding can be done later. **Incrementally** in an **Agile** manner, By **Multiple** contributors. 

The source of all confusion, rests on the fact that, I have named the **content** container /component in the list above, as **greensock**. 

I can always change it to **content**, so there is no confusion.

Now, This PR, is only about, 

1. Defining three `sections`, the react way (as in sampleContainer/sampleComponent and how they are bound and laid out on page) BY:
a. Defining a `react container`  and `react component` for these three sections that we need
b. Defining a `react container to react component` binding, **so we retain**  a `react` way of logically seperating concerns 

4. We are then finally, defining the `app` as a `flexbox`, and making the sections responsive.


###What we are not doing

1. Deciding how we animate
2. Deciding how data is stored
3. Deciding / Building the intermediate / final UI 

Which can be done either by the lot of us discussing this PR, or by others, at their own times. 
I am open to finishing the complete UI too, but this, is just to define the emptyness and organize it for now.

skyshader replied:
Yes, @retrocausal it is clear to me what you are trying to do. But `ContentContainer` doesn't sound reasonable to me. We can have a `ContentComponent` though which wraps everything inside it into `<section>` tag with some default styles for content. But mostly we would have a `QuizContainer` which will contain `QuestionComponent` and different `OptionComponent`.

It's not necessary to always have a container for every single component. But a component should only be included in a container. Like above quiz is handling question and options.

Let's do this:
1. Keep the Header (Can be used to show the hamburger menu and the score and other things)
2. Create a Sidebar, which will slide-in when we tap on hamburger menu.
3. Let's have a Navigation inside the Sidebar.
4. Keep the Footer and we will see if we need it. If there is just static data and we don't really need any state in this, we can skip FooterContainer and just keep the component.

This summarises what the app can actually use. Get in touch with me if you have any doubts or queries :)

retrocausal replied:
That clears things up! will make these changes, and commit. I just wanted to have one exhaustive comment, and clarify just in case I was infusing confusion in to the unusually long thread on this PR :laughing: 
Thanks Shaan!

skyshader commented on client/src/components/GreensockComponent/GreensockComponent.js:
```diff
@@ -0,0 +1,12 @@
+};
+
+GreensockComponent.propTypes = {
+  message: PropTypes.string
```
Let's add a default prop if we are not marking prop types as required.

retrocausal replied:
currently, the placeholder text we are using to depict the mockup, is defined as a default property called `placeHolder`. 
the property `message`, is from the react app bootstrapped/generated code, which is a part of the redux store. I have not deleted the generated properties. can do so if need be.

What I did was, copy the sample container over, and create three containers similar to it. Hence, the new containers have the react generated defaults too. I was not sure, if we need them or not, but the generated code, serves as an example for using redux stores. 
so left the message and sample properties untouched.

skyshader replied:
Okay. We can ignore this for now. This was added by me though the sample container and stuff. But it's better to add something like this:
```js
GreensockComponent.defaultProps = {
  message: '',
};
```
when we don't use `isRequired` for any proptype in props validation.

skyshader reviewed (commented):
Also, as a general comment. Do we really have a footer component in our UI mocks? I don't think so.

Here is an example: https://github.com/VarunDevPro/gtq-designs/blob/master/03-question.png

skyshader commented:
Let's have a sidebar as well `aside`. Because we do have a menu listing in sidebar.

retrocausal committed:
Merge branch 'udevelop' into develop

retrocausal committed:
Merge branch 'udevelop' into develop

retrocausal committed:
New scaffolds for UI overhauling
 - Remove components and containers for greensock
 - Add a Content component
 - Add a Quiz component
 - Add a Quiz container
 - Let the home page only render
    - HeaderComponent
    - QuizContainer
 - The QuizContainer needs to define render props
    - And bind it to the QuizComponent

twishasaraiya commented:
Thanks @retrocausal for the contribution :+1:  . This looks good to me but I am not still sure about footer component but let's keep it. Also let's wait for what @skyshader and @anantoghosh has to say.

retrocausal commented:
@twishasaraiya , I got a picture of what you guys are trying to achieve. Regarding the footer, I have not included it on the home page, but have a component ready. In case we need it. I will be adding the off canvas sidebar soon. Maybe Monday. we could then look at styling the home page which would be the landing page by this weekend. If you could decide what needs to be on the header, I can add it too.

twishasaraiya commented:
Closing this request since we have already moved ahead in the project.
